Fix cube fitting compatibility with unit conversion

rosteen opened the pull request:
Opening as draft since I still need to work on the unit conversion part, but I wanted to see the diff and remind myself what I'd changed so far.

cshanahan1 commented:
Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:

`cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)`

cshanahan1 reviewed (commented):
Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:

cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)

rosteen commented:
> Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:
> 
> cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)

Hm, the model fit plots fine with the example notebook data, but now that you mention it I don't see table results. I'll look into it on Monday.

rosteen commented:
> Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:
> 
> cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)

From a comment in the code: `# cube fits are currently unsupported in tables` - I think that's out of scope for this. I'll investigate the test fixture not plotting.

rosteen commented:
https://github.com/user-attachments/assets/9bfe567d-39ae-47f4-abbd-a1bc35b3f9a4

@cshanahan1  hm, I see the model plot fine with that cube (I fit a constant to it). Are you sure you didn't need to go into the data menu and add the model to the viewer? Or did it fail to appear with the `Plot in Viewer` option toggled on?

 I do see an error trying to add a Linear model component, trying to fix that...

rosteen commented:
Note that this merging this PR on top of #3156 resolves the test failures (other than the known axis label failure there).

cshanahan1 commented:
![Screenshot 2024-09-16 at 1 49 02 PM](https://github.com/user-attachments/assets/4e6dca92-d735-4c9f-b18b-02d8d8a80cef)
still seeing some out of sync units. I loaded a Jy/sr cube, changed the flux unit to MJy, and the model fitting plugin shows Jy as the flux unit. This happened when MJy was set before creating the component for the first time, so I shouldn't have to re-estimate parameters, but even when doing that the incorrect unit remains

camipacifici commented:
Linear fitting to cube or single spectrum does not break anymore!
In this PR, I see the icon to open the model fitting table in a different window, but the link goes to a 404 window.

rosteen commented:
> still seeing some out of sync units. I loaded a Jy/sr cube, changed the flux unit to MJy, and the model fitting plugin shows Jy as the flux unit. This happened when MJy was set before creating the component for the first time, so I shouldn't have to re-estimate parameters, but even when doing that the incorrect unit remains


Fixed, toggling the cube fit was triggering the dataset selected logic, which was resetting the units. I removed all the unit setting from there in favor of setting it when the first component is initialized (and whenever user changes units or toggles cube fit).

kecnry commented on jdaviz/configs/default/plugins/model_fitting/model_fitting.py:
```diff
@@ -470,6 +465,17 @@ def _initialize_model_component(self, model_comp, comp_label, poly_order=None):
+        # we listen for display unit changes
+        if (self._units is None or self._units == {} or 'x' not in self._units or
+                'y' not in self._units):
+            uc = self.app._jdaviz_helper.plugins['Unit Conversion']
+            self._units['x'] = uc.spectral_unit.selected
+            if self.cube_fit:
+                self._units['y'] = uc._obj.sb_unit_selected
+            else:
+                self._units['y'] = uc.flux_unit.selected
```
eventually this probably will be safe, but might need to use `app._get_display_units` for now to get the fallbacks when unit conversion does not exist

kecnry replied:
for non-cube fit, this also needs to listen to `spectral_y_type` (`app._get_display_units('spectral_y')` handles that logic as well).

cshanahan1 commented:
I see it now respecting flux unit selection when i change between Jy/MJy, but breaks on erg / (Angstrom s cm2 sr)

kecnry commented on jdaviz/configs/default/plugins/model_fitting/model_fitting.py:
```diff
@@ -309,6 +299,30 @@ def _warn_if_no_equation(self):
+            self.dataset.remove_filter('is_flux_cube')
+
+        self.dataset._clear_cache()
+        self.reestimate_model_parameters()
```
If we remove `reestimate_model_parameters` from here does this still work (just sometimes requiring the user to hit the button)?

I wonder if its ever useful to keep the estimates from the non cube fit as the starting point (especially if there is no unit conflict)?  Or if the user manually entered there own values but then they were changed by toggling cube fit, would that be unexpected/disruptive?

rosteen replied:
Updated to only reestimate if `spectral_y` and `sb` have different units.

rosteen commented:
Based on offline feedback from @camipacifici I changed the cube fit toggle to display the incompatible units warning, prompting the user to click to reestimate the parameters rather than doing it automatically. A future improvement will be to automatically convert using the scale factor if possible.

rosteen commented:
Oh right - the test coverage is going to look bad because there are some getting skipped that need #3156.

cshanahan1 commented:
> display the incompatible units warning, prompting the user to click to reestimate the parameters rather than doing it automatically

I don't see this working. Using cubeviz example notebook, If I start in MJy and add a constant component and fit and plot a cube fit, then change the unit, the component is still in MJy with no warning. Interestingly, the fit plotted in the viewer after the unit change has the correct units, but the model component displayed does not

cshanahan1 commented:
Could you add a test that tests cube fitting with unit conversion(s)? Maybe something where you change from Jy > MJy, and Jy > er/s/cm2/A ?

rosteen commented on jdaviz/configs/default/plugins/model_fitting/model_fitting.py:
```diff
@@ -309,6 +299,36 @@ def _warn_if_no_equation(self):
+        self._update_viewer_filters(event=event)
+
+        sb_unit = self.app._get_display_unit('sb')
+        spectral_y_unit = self.app._get_display_unit('spectral_y')
```
I'm using it a few lines down to check to see if the collapsed spectrum is in flux or surface brightness, to see if the parameter units will be compatible when toggling `cube_fit`.

rosteen commented:
@cshanahan1 I added a test, but I won't be able to test changing the unit to `erg / (Angstrom s cm2)` until the contour bug is fixed.

cshanahan1 commented:
Tested this one final time and couldn't break it.  Talked on slack with ricky about a few points of confusion i had and he cleared those up so I'll approve, looks good!

cshanahan1 commented on jdaviz/configs/default/plugins/model_fitting/model_fitting.py:
```diff
@@ -930,7 +984,8 @@ def _fit_model_to_cube(self, add_data):
                 run_fitter=True,
-                window=None
+                window=None,
+                n_cpu=1  # Remove this after debugging!
```
pssst

gibsongreen reviewed (approved):
Did one last look over and round of testing, all looks good to me and works as I'd expect!
